tracer/src/Datadog.Trace/SpanExtensions.cs

```diff
@@ -30,12 +36,31 @@ public static void SetUser(this ISpan span, UserDetails userDetails)
             }
 
             var localRootSpan = span;
+            TraceContext traceContext = null;
             if (span is Span spanClass)
             {
-                localRootSpan = spanClass.Context.TraceContext?.RootSpan ?? span;
+                traceContext = spanClass.Context.TraceContext;
+                localRootSpan = traceContext?.RootSpan ?? span;
             }
 
-            localRootSpan.SetTag(Tags.User.Id, userDetails.Id);
+            if (userDetails.PropagateId)
+            {
+                var base64UserId = Convert.ToBase64String(Encoding.UTF8.GetBytes(userDetails.Id));
+
+                var propagationHeaderMaxLength =
+                    traceContext?.Tracer.Settings.TagPropagationHeaderMaxLength ??
+                        TagPropagation.OutgoingPropagationHeaderMaxLength;
+                if (base64UserId.Length > propagationHeaderMaxLength)
+                {
+                    Log.Warning<string, int>("{Id} is {IdLength} bytes long, which is longer than the configured max length of {MaxLength}", userDetails.Id, base64UserId.Length, propagationHeaderMaxLength);
+                }
+
+                localRootSpan.SetTag(TagPropagation.PropagatedTagPrefix + Tags.User.Id, base64UserId);
```

Comment:
nit: I had to check if these were constants (and so shouldn't allocate) - might be nice to make that explicit? e.g. 
```suggestion
                const string propagatedUserIdTag = TagPropagation.PropagatedTagPrefix + Tags.User.Id;
                localRootSpan.SetTag(propagatedUserIdTag, base64UserId);
```